politeiad/backendv2/tstorebe/plugins/pi/proposalstatus.go

```diff
@@ -0,0 +1,191 @@
+			return "", err
+		}
+	}
+
+	// Determine the proposal status
+	propStatus, err = proposalStatus(recordState, recordStatus, voteStatus,
+		voteMetadata, billingStatuses)
+	if err != nil {
+		return "", nil
+	}
+
+	// Cache the results
+	p.statuses.set(tokenStr, statusEntry{
+		propStatus:   propStatus,
+		recordState:  recordState,
+		recordStatus: recordStatus,
+		voteStatus:   voteStatus,
+		voteMetadata: voteMetadata,
+	})
+
+	return propStatus, nil
+}
+
+// statusIsFinal returns whether the proposal status is a final status and
+// cannot be changed any further.
+func statusIsFinal(s pi.PropStatusT) bool {
+	switch s {
+	case pi.PropStatusUnvettedAbandoned, pi.PropStatusUnvettedCensored,
+		pi.PropStatusAbandoned, pi.PropStatusCensored, pi.PropStatusApproved,
+		pi.PropStatusRejected:
+		return true
+	default:
+		return false
+	}
+}
+
+// statusRequiresRecord returns whether the proposal status requires the record
+// to be retrieved from the backend. This is necessary when the proposal is in
+// a part of the proposal lifecycle that still allows changes to the underlying
+// record data. For example, an unvetted proposal may still have it's record
+// metadata or vote metadata altered, but a proposal with the status of active
+// cannot.
+func statusRequiresRecord(s pi.PropStatusT) bool {
+	if statusIsFinal(s) {
+		// The status is final and cannot be changed
+		// any further, which means the record data
+		// is not required.
+		return false
+	}
+
+	switch s {
+	case pi.PropStatusVoteStarted, pi.PropStatusActive,
+		pi.PropStatusCompleted, pi.PropStatusClosed:
+		// The record cannot be changed any further for
+		// these statuses.
+		return false
+
+	case pi.PropStatusUnvetted, pi.PropStatusUnderReview,
+		pi.PropStatusVoteAuthorized:
+		// The record can still change for these statuses.
+		return true
+
+	default:
+		// Defaulting to true is the conservative default
+		// since it will force the record to be retrieved
+		// for unhandled cases.
+		return true
+	}
+}
+
+// requiresVoteSummary returns whether the proposal requires the vote summary
+// to be retrieved to determine the proposal status on runtime. This is
+// necessary when the proposal is in a stage where the vote status can still
+// change.
+func requiresVoteSummary(s pi.PropStatusT, state backend.StateT, status backend.StatusT) bool {
+	if statusIsFinal(s) {
+		// The status is final and cannot be changed
+		// any further, which means the vote summary
+		// is not required.
+		return false
+	}
+
+	switch s {
+	case pi.PropStatusActive, pi.PropStatusClosed, pi.PropStatusCompleted:
+		// The vote result is known no need to fetch
+		return false
+
+	default:
+		// If proposal status is unknown, not final or vote was not finished yet,
+		// vote summary is required only if proposal is vetted and public.
+		return state == backend.StateVetted && status == backend.StatusPublic
+	}
+}
+
+// requiresBillingStatuses returns whether the proposal metadata and vote
```

Comment:
// requiresBillingStatuses returns whether the proposal requires the billing
// status changes to be retrieved to determine the...